mage/main.go

```diff
@@ -456,9 +460,64 @@ func Invoke(inv Invocation) int {
 		return 0
 	}
 
+	if noCache, err := strconv.ParseBool(os.Getenv("MAGE_NOCACHE")); err == nil && noCache {
+		defer os.Remove(exePath)
+	}
+
+	autoCleanup, err := strconv.ParseBool(os.Getenv("MAGEFILE_AUTOCLEANUP"))
+	if err != nil {
+		autoCleanup = true
+	}
+	if autoCleanup {
+		defer cleanCache(inv.CacheDir)
+	}
+
 	return RunCompiled(inv, exePath, errlog)
 }
 
+func cleanCache(cacheDir string) error {
+	maxCacheSize, _ := strconv.ParseInt(os.Getenv("MAGEFILE_MAX_CACHE_SIZE"), 10, 64)
+	maxCacheAge, _ := strconv.ParseInt(os.Getenv("MAGEFILE_MAX_CACHE_AGE"), 10, 64)
+
+	if maxCacheSize == 0 {
+		maxCacheSize = defaultMaxCacheSize
+	}
+
+	if maxCacheAge == 0 {
+		maxCacheAge = defaultMaxCacheAge
+	}
+
+	files, err := ioutil.ReadDir(cacheDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return err
+	}
+	sort.Slice(files, func(i, j int) bool {
+		return files[i].ModTime().After(files[j].ModTime())
+	})
+
+	timeLimit := time.Now().Local()
+	maxCacheDays := time.Duration(-maxCacheAge*24) * time.Hour
+	timeLimit = timeLimit.Add(maxCacheDays)
+
+	var totalSize int64 = 0
+	for _, f := range files {
+		if f.IsDir() {
+			continue
+		}
+		totalSize += f.Size()
+		if maxCacheAge > 0 && timeLimit.After(f.ModTime()) {
+			os.Remove(filepath.Join(cacheDir, f.Name()))
+		}
+		if maxCacheSize > 0 && totalSize > maxCacheSize {
+			os.Remove(filepath.Join(cacheDir, f.Name()))
+		}
```

Comment:
The computed size of the cache shouldn't include files that are deleted -- it'd cause excess files to be removed.

```suggestion
		if maxCacheAge > 0 && timeLimit.After(f.ModTime()) {
			os.Remove(filepath.Join(cacheDir, f.Name()))
		}
		if maxCacheSize > 0 && totalSize + f.Size() > maxCacheSize {
			os.Remove(filepath.Join(cacheDir, f.Name()))
		}
		totalSize += f.Size()
```

Reply:
Very good point, this is failing when you combine the 2 policies at the same time. Thanks! I'll add a test for that too.

Reply:
Actually, I realized that is not the case, there are ordered by modification date, so... if you already remove one file because the modification date, you are going to remove every other file after that, so... doesn't matter if the size is affected or not. On the other hand, if the size is the one triggering the remove, you already surpassed the size, so the age is not relevant anymore... I think the code works well, anyway I going to do it as an `else if` to avoid duplicated `os.Reomve`.

Reply:
re the ordering: Consider a size limit of 10 and a sequence of cached files of sizes 2,9,1,1,1 (after timestamp ordering)

Your algorithm will delete everything except the first one. But in reality only the "9" needs to be deleted.

In general, anything that is deleted by definition does not contribute to the total size :)

Reply:
Interesting point. I would expect that size of files are going to be pretty homogeneous, but its is true that you can have dependencies and in consequence generate pretty big binaries. Let me try to create a test case for that. I don't think the current policy is necessarily wrong (We can assume that everything that is after certain size, gets removed, and always by date), but I agree that can be better in that case.